base/compiler/abstractinterpretation.jl

```diff
@@ -1232,6 +1232,27 @@ function abstract_call_builtin(interp::AbstractInterpreter, f::Builtin, (; fargs
                 end
                 return Conditional(aty.var, ifty, elty)
             end
+        elseif f === isdefined
+            uty = argtypes[2]
+
+            if isa(uty, Union)
+                fld = argtypes[3]
+                if isa(fld, Const)
+                    fld = fld.val
+                end
```

Comment:
`isdefined_tfunc` seems to return just `Bool` for some of these test cases, in which case we can't determine what to do with `vtype`/`elsetype` --- I would suggest still using `hasfield` but just bailing out if `fld` isn't a symbol, since we wouldn't be able to do any narrowing in that case anyway

Reply:
We'd perhaps still be able to filter out `nothing` and other objects without any fields?

Reply:
Ok, I can do that instead of just bailing out. Just not sure how we can use `isdefined_tfunc` here when it just gives `Bool` for all the test case examples

Reply:
See latest update